core/src/main/java/org/jdbi/v3/core/qualifier/QualifiedType.java

```diff
@@ -39,7 +38,8 @@
 @Beta
 public final class QualifiedType<T> {
     private final Type type;
-    private final Set<Annotation> qualifiers;
+    // null for no qualifiers, Annotation itself for one qualifier, Set<Annotation> for many
+    private final Object qualifiers;
```

Comment:
Could it be interesting to encapsulate this variable type into its own class, to separate that concern from the rest of QualifiedType? I'd have to give it a try to be sure but I have a feeling it would make the current spaghetti of constructors and factories calling each other simpler.